bindings/python/pymongocrypt/mongocrypt.py

```diff
@@ -436,24 +449,36 @@ def __init__(self, ctx, kms_provider, opts):
         """
         super(DataKeyContext, self).__init__(ctx)
         try:
-            if kms_provider == 'aws':
+            if kms_provider in ['aws', 'gcp', 'azure']:
                 if opts is None or opts.master_key is None:
                     raise ValueError(
-                        'master_key is required for kms_provider: "aws"')
-                if ('region' not in opts.master_key or
-                        'key' not in opts.master_key):
-                    raise ValueError(
-                        'master_key must include "region" and "key" for '
-                        'kms_provider: "aws"')
-                region = str_to_bytes(opts.master_key['region'])
-                key = str_to_bytes(opts.master_key['key'])
-                if not lib.mongocrypt_ctx_setopt_masterkey_aws(
-                        ctx, region, len(region), key, len(key)):
-                    self._raise_from_status()
-                if 'endpoint' in opts.master_key:
-                    endpoint = str_to_bytes(opts.master_key['endpoint'])
-                    if not lib.mongocrypt_ctx_setopt_masterkey_aws_endpoint(
-                            ctx, endpoint, len(endpoint)):
+                        'master_key is required for kms_provider: "%s"' % (
+                            kms_provider,))
+                if kms_provider == 'aws':
+                    if ('region' not in opts.master_key or
+                            'key' not in opts.master_key):
+                        raise ValueError(
+                            'master_key must include "region" and "key" for '
+                            'kms_provider: "aws"')
+                elif kms_provider == 'azure':
+                    if ('keyName' not in opts.master_key or
+                            'keyVaultEndpoint' not in opts.master_key):
+                        raise ValueError(
+                            'master key must include "keyName" and '
+                            '"keyVaultEndpoint" for kms_provider: "azure"')
+                else:   # kms_provider == 'gcp'
+                    if ('projectId' not in opts.master_key or
+                            'location' not in opts.master_key or
+                            'keyRing' not in opts.master_key or
+                            'keyName' not in opts.master_key):
+                        raise ValueError(
+                            'master key must include "projectId", "location",'
+                            '"keyRing", and "keyName" for kms_provider: "gcp"')
+                master_key = opts.master_key.copy()
+                master_key['provider'] = kms_provider
+                with MongoCryptBinaryIn(bson_encode(master_key)) as mkey:
+                    if not lib.mongocrypt_ctx_setopt_key_encryption_key(
+                            ctx, mkey.bin):
                         self._raise_from_status()
             elif kms_provider == 'local':
```

Comment:
I wasn't able to figure out how to use `mongocrypt_ctx_setopt_key_encryption_key` to set the KEK for `local` provider. What should be passed as the second argument to this method for `local` @kevinAlbs ? Should it be Null?

Reply:
It should be the document `{"provider": "local"}`. https://github.com/mongodb/libmongocrypt/blob/master/src/mongocrypt-kek.c#L21 lists what this expects to parse.

Reply:
Pretty sure you just need to pass `bson_encode({'provider': 'local'})`. See https://github.com/mongodb/libmongocrypt/blob/509ae2ec26000b9351e322d38b06b3a0d1aa873f/test/util/csfle.c#L178-L187